daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala

```diff
@@ -103,6 +104,9 @@ class VariableInstance private (
 
   lazy val defaultValueExpr = defaultValueExprArg
 
+  // This represents the default value at the start of processing, provided by
```

Comment:
There are some instances where we are using the defaultValueExpression in places without access to a variable runtime data. For example, we check to see if a variable has a defaultValueExpression defined when resetting the instance to determine the correct state to reset to. We also evaluate the expressions in forceExpressionEvaluations without access to a VRD.

It may be possible to restructure these things, but I'm not sure its worth the effort.

Reply:
Yeah, I think you're right that there are places where we only have access to the instance.

That said, I think anything that can simplify this variable stuff is worth the effort. It's taken me multiple reviews to get an idea of how it all works, and I still don't understand it all (that isn't saying anything about yoru code, this stuff is just complicated).

For example, I noticed that each ``VariableInstance`` already has a pointer to its VariableRuntimeData, so could we remove these defaultValueExpr/Arg members and access the expression on the RuntimeData when needed?

Also, I noticed in the VRD that it stores state and value depending on if the if there's a defaultValue expression, and then that is used to initialize the VariableInstance. This feels unnecessary and wrong to me. We shouldnt' be creating a VariableInstance with some state based on whether or not it has a defaultValue, since things like external values also play a role. It seems like we should always create an empty VariableInstance with no value/state, and the all the new you've added should be setting default value/state as appropriate. In fact, my guess is this is where the value=NoValue but state=VariableDefined comes from.

All that said, I wonder if instead, a VariableInstance should take no parameters except for VariableRuntimeData. So the value, state, priorValue, priorState, and defaultValue become member variables. When we create a VariableInstance, it's value is initialize to None and state to VariableUndefined. Then all the logic you've added will get triggered during parse to figure out default values and set state appropriately.

Reply:
The VariableInstance constructor takes only a VRD now. So this thread seems to be resolved.